src/hotspot/cpu/aarch64/immediate_aarch64.cpp

```diff
@@ -148,11 +161,71 @@ uint64_t replicate(uint64_t bits, int nbits, int count)
   return result;
 }
 
-// this function writes the supplied bimm reference and returns a
-// boolean to indicate success (1) or fail (0) because an illegal
-// encoding must be treated as an UNALLOC instruction
+// construct a 64 bit immediate value for a logical immediate operation
+//
+// SPEC:
+//
+// {(0,_), (1, uint64)} = expandLogicalImmediate(immN, immr, imms)
+//
+// For valid combinations of immN, immr and imms, this function
+// replicates a derived bit string of width 2^k into a 64 bit result
+// and returns 1.
+//
+// for invalid combinations it fails and returns 0
+//
+// - immN and imms together define
+//
+//    1) the size of the bit field (2^k) to be replicated
+//
+//    2) how many bits to set in that field, starting from bit 0
+//
+// - immr defines a right rotation on the 2^k bit field
+//
+// example:
+//
+// valid outputs will look like:
+//
+//   11001111_11001111_11001111_11001111_11001111_11001111_11001111_11001111
+//
+// which corresponds to the inputs
+//
+//   immN = 0, imms = 110101, immr = 000010
+//
+// For these inputs k = 3,  2^k = 8, set bits = 6, rotation = 2
+//
+// validity checks:
+//
+// when immN is 1 then k == 6 and immr/imms are masked to 6 bit
+// integers
+//
+// when immN is 0 then k is the index of the first 0 bit in imms and
+// immr/imms are masked to k-bit integers (i.e. any leading 1s and the
+// first 0 in imms determine dead bits of imms/immr)
+//
+// if (pre-masking) immr >= 2^k then fail and return 0 (this is a
+// uniqueness constraint that ensures each output bit string is only
+// generated by one valid combination of immN, imms and immr).
+//
+// if k == 0 the fail and return 0. It is not possible to construct a
+// mask from an zero width bit string.
+//
+// If imms == all 1s (modulo 2^k) then fail and return 0.  Note that
+// this means that 0 <= imms < 2^k - 1, i.e. 0 < imms + 1 < 2^k. This
+// ensures that an all 0s or all 1s bit pattern is never generated.
+//
+// bit field construction:
```

Comment:
The flow in this section is odd. I think the text should flow as: "This is what we intend to do; this is how we construct it is done; this is the example of the generated output; given all above; here are the invariants that are enforced".

Reply:
Yeah, I agree, only I found it tricky to describe the algorithm without describing the constraints that determine how it proceeds. I have pushed a revised version to address that.